File: contests/Aoc/2025/2/solution.go
package main

import (
	"flag"
	"fmt"
	"log"
	"math"
	"os"
	"sort"
	"strconv"
	"strings"
)

func main() {
	inputPath := flag.String("input", "input.txt", "path to input file")
	partTwo := flag.Bool("part2", false, "use part two rules (repeat >= 2 times)")
	flag.Parse()

	// Allow a positional override for the input path to stay convenient.
	if flag.NArg() > 0 {
		*inputPath = flag.Arg(0)
	}

	data, err := os.ReadFile(*inputPath)
	if err != nil {
		log.Fatalf("read %s: %v", *inputPath, err)
	}

	line := strings.TrimSpace(string(data))
	if line == "" {
		fmt.Println("0")
		return
	}

	fmt.Println(solveLine(line, *partTwo))
}

func solveLine(line string, partTwo bool) int64 {
	ranges, maxEnd := parseRanges(line)
	maxRepeat := 2
	if partTwo {
		maxRepeat = 0 // 0 means unlimited (>=2)
	}

	repeats := generateRepeatedNumbers(maxEnd, maxRepeat)
	prefix := make([]int64, len(repeats)+1)
	for i, v := range repeats {
		prefix[i+1] = prefix[i] + v
	}

	var total int64
	for _, rg := range ranges {
		l := sort.Search(len(repeats), func(i int) bool { return repeats[i] >= rg.start })
		r := sort.Search(len(repeats), func(i int) bool { return repeats[i] > rg.end })
		total += prefix[r] - prefix[l]
	}
	return total
}

func pow10Safe(n int) (int64, bool) {
	var r int64 = 1
	for i := 0; i < n; i++ {
		if r > math.MaxInt64/10 {
			return 0, false
		}
		r *= 10
	}
	return r, true
}

func powInt(base int64, exp int) (int64, bool) {
	r := int64(1)
	for i := 0; i < exp; i++ {
		if willOverflowMul(r, base) {
			return 0, false
		}
		r *= base
	}
	return r, true
}

func willOverflowMul(a, b int64) bool {
	if a == 0 || b == 0 {
		return false
	}
	if a == 1 || b == 1 {
		return false
	}
	if a == -1 {
		return b == math.MinInt64
	}
	if b == -1 {
		return a == math.MinInt64
	}
	// Ensure |a| <= |b|
	if a < 0 {
		if b < 0 {
			if a < b {
				a, b = b, a
			}
		} else {
			if -a > b {
				a, b = b, a
			}
		}
	} else if b < 0 && a > -b {
		a, b = b, a
	}

	return a > math.MaxInt64/b || a < math.MinInt64/b
}

func ceilDiv(a, b int64) int64 {
	if b == 0 {
		return 0
	}
	q := a / b
	if a%b != 0 {
		q++
	}
	return q
}

type numRange struct {
	start int64
	end   int64
}

func parseRanges(line string) ([]numRange, int64) {
	var rs []numRange
	var maxEnd int64
	for _, part := range strings.Split(line, ",") {
		part = strings.TrimSpace(part)
		if part == "" {
			continue
		}
		bounds := strings.Split(part, "-")
		if len(bounds) != 2 {
			log.Fatalf("invalid range %q", part)
		}
		start, err := strconv.ParseInt(bounds[0], 10, 64)
		if err != nil {
			log.Fatalf("invalid start %q: %v", bounds[0], err)
		}
		end, err := strconv.ParseInt(bounds[1], 10, 64)
		if err != nil {
			log.Fatalf("invalid end %q: %v", bounds[1], err)
		}
		if start > end {
			start, end = end, start
		}
		if end > maxEnd {
			maxEnd = end
		}
		rs = append(rs, numRange{start: start, end: end})
	}
	return rs, maxEnd
}

func generateRepeatedNumbers(limit int64, maxRepeat int) []int64 {
	set := make(map[int64]struct{})
	for lenSeq := 1; ; lenSeq++ {
		pow, ok := pow10Safe(lenSeq)
		if !ok {
			break
		}
		minSeq := pow / 10 // smallest sequence without leading zero
		if minSeq == 0 {
			continue
		}
		maxSeq := pow - 1

		for repeat := 2; ; repeat++ {
			if maxRepeat > 0 && repeat > maxRepeat {
				break
			}

			powK, ok := powInt(pow, repeat)
			if !ok {
				break
			}
			geom := (powK - 1) / (pow - 1) // 1 + pow + pow^2 + ... + pow^(repeat-1)
			minVal := minSeq * geom
			if minVal < 0 || minVal > limit {
				break // further repeats will only get larger
			}
			endSeq := maxSeq
			maxVal := maxSeq * geom
			if maxVal < 0 {
				break
			}
			if maxVal > limit {
				endSeq = limit / geom
			}
			for seq := minSeq; seq <= endSeq; seq++ {
				val := seq * geom
				if val <= limit {
					set[val] = struct{}{}
				}
			}
		}
	}

	nums := make([]int64, 0, len(set))
	for v := range set {
		nums = append(nums, v)
	}
	sort.Slice(nums, func(i, j int) bool { return nums[i] < nums[j] })
	return nums
}

File: contests/Aoc/2025/2/solution.py
import argparse
import bisect
from typing import List, Tuple


def parse_ranges(line: str) -> Tuple[List[Tuple[int, int]], int]:
    ranges: List[Tuple[int, int]] = []
    max_end = 0
    for part in line.split(","):
        part = part.strip()
        if not part:
            continue
        if "-" not in part:
            raise ValueError(f"invalid range {part!r}")
        start_s, end_s = part.split("-", 1)
        start = int(start_s)
        end = int(end_s)
        if start > end:
            start, end = end, start
        ranges.append((start, end))
        if end > max_end:
            max_end = end
    return ranges, max_end


def generate_repeated_numbers(limit: int, max_repeat: int) -> List[int]:
    # Generate all numbers up to limit that are formed by repeating a digit
    # sequence k times, k >= 2. If max_repeat == 0, allow any k >= 2.
    seen = set()
    len_seq = 1
    while True:
        pow10 = 10 ** len_seq
        min_seq = pow10 // 10
        if min_seq == 0:
            len_seq += 1
            continue

        any_added = False
        repeat = 2
        while True:
            if max_repeat and repeat > max_repeat:
                break

            geom = (pow10 ** repeat - 1) // (pow10 - 1)  # 1 + pow + ... + pow^(k-1)
            min_val = min_seq * geom
            if min_val > limit:
                break

            max_seq = pow10 - 1
            end_seq = min(max_seq, limit // geom)
            if end_seq >= min_seq:
                any_added = True
                for seq in range(min_seq, end_seq + 1):
                    val = seq * geom
                    if val <= limit:
                        seen.add(val)
            repeat += 1

        # If even the smallest repeated number for this len_seq exceeds limit, stop.
        if not any_added and (min_seq * ((pow10 * pow10 - 1) // (pow10 - 1)) > limit):
            break
        len_seq += 1

    nums = sorted(seen)
    return nums


def build_prefix(nums: List[int]) -> List[int]:
    prefix = [0]
    for v in nums:
        prefix.append(prefix[-1] + v)
    return prefix


def solve(line: str, part2: bool) -> int:
    ranges, max_end = parse_ranges(line)
    max_repeat = 0 if part2 else 2
    nums = generate_repeated_numbers(max_end, max_repeat)
    prefix = build_prefix(nums)

    total = 0
    for start, end in ranges:
        l = bisect.bisect_left(nums, start)
        r = bisect.bisect_right(nums, end)
        total += prefix[r] - prefix[l]
    return total


def main() -> None:
    parser = argparse.ArgumentParser(description="Sum repeated-pattern IDs in ranges.")
    parser.add_argument("path", nargs="?", default="input.txt", help="input file path")
    parser.add_argument("--part2", action="store_true", help="use part two rules (repeat >= 2 times)")
    args = parser.parse_args()

    with open(args.path, "r", encoding="utf-8") as f:
        line = f.read().strip()
    if not line:
        print(0)
        return

    result = solve(line, args.part2)
    print(result)


if __name__ == "__main__":
    main()
